src/key_info_managers/sqlite_manager/mod.rs

```diff
@@ -0,0 +1,597 @@
+            )
+            ",
+            [],
+        )?;
+
+        let mut stmt = conn.prepare("SELECT * FROM key_mapping")?;
+        let key_iter = stmt.query_map([], |row| {
+            let key_identity = KeyIdentity::new(
+                ApplicationIdentity::new(
+                    row.get("application_name")?,
+                    i64_to_auth_type(row.get("authenticator_id")?).map_err(|e| {
+                        format_error!("Failed to get AuthType from authenticator_id.", e);
+                        let error = Box::new(Error::new(ErrorKind::InvalidData, e));
+                        RusqliteError::FromSqlConversionFailure(64, Integer, error)
+                    })?,
+                ),
+                ProviderIdentity::new(row.get("provider_uuid")?, row.get("provider_name")?),
+                row.get("key_name")?,
+            );
+            let key_info_blob: Vec<u8> = row.get("key_info")?;
+
+            // TODO: Change this to (protobuf?) version once format has been decided.
+            let key_info = bincode::deserialize(&key_info_blob[..]).map_err(|e| {
+                format_error!("Error deserializing key info", e);
+                RusqliteError::FromSqlConversionFailure(key_info_blob.len(), Blob, e)
+            })?;
+
+            Ok(KeyInfoRecord {
+                key_identity,
+                key_info,
+            })
+        })?;
+
+        // Add keys to key_store cache
+        for key_info_record in key_iter {
+            let key_info_record = key_info_record?;
+            let _ = key_store.insert(key_info_record.key_identity, key_info_record.key_info);
+        }
+
+        if !crate::utils::GlobalConfig::log_error_details() {
+            info!(
+                "SQLiteKeyInfoManager - Found {} key info mapping records",
+                key_store.len()
+            );
+        }
+
+        Ok(SQLiteKeyInfoManager {
+            key_store,
+            database_path,
+        })
+    }
+
+    /// Saves the KeyIdentity and KeyInfo to the database.
+    /// Inserts a new record to the database `key_mapping` table.
+    fn save_mapping(
+        &self,
+        key_identity: &KeyIdentity,
+        key_info: &KeyInfo,
+    ) -> rusqlite::Result<(), RusqliteError> {
+        let conn = Connection::open(&self.database_path)?;
+
+        // TODO: Change this to (protobuf?) version once format has been decided.
+        let key_info_blob = bincode::serialize(&key_info).map_err(|e| {
+            format_error!("Error serializing key info", e);
+            RusqliteError::ToSqlConversionFailure(e)
+        })?;
+
+        // Insert the new key mapping, if a record does not exist.
+        // If one does exist, update the existing record.
+        let _ = conn.execute(
+            "
+            INSERT OR IGNORE INTO
+                `key_mapping`
+                (`authenticator_id`, `application_name`, `provider_uuid`, `provider_name`, `key_name`, `key_info`, `key_info_version`)
```

Comment:
My current plan is before these queries are run, check whether any keys need to be migrated to the current version. If they do need upgrading; either throw an error if we end up deciding on a standalone migration program (e.g. "please use the key info migration program to upgrade your keys"), or automatically upgrade the keys within the parsec service if this is the approach we decide. If all keys are up to date it will reach this logic and load the KIM mappings.

TLDR; should never hit these load statements with an incorrect `key_info_version`.

Reply:
Cool, thanks for the info! 👍🏻